// angular-translator/src/TranslateComponent.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { TranslateLogHandler } from "./TranslateLogHandler";
import { Translator } from "./Translator";
import { TranslatorContainer } from "./TranslatorContainer";
import { Component, Input } from "@angular/core";
var TranslateComponent = /** @class */ (function () {
    function TranslateComponent(translator, translatorContainer, logHandler) {
        var _this = this;
        this.translator = translator;
        this.translatorContainer = translatorContainer;
        this.logHandler = logHandler;
        this.translation = "";
        this._params = {};
        this.subscription = this.translator.languageChanged.subscribe(function () {
            _this.startTranslation();
        });
    }
    Object.defineProperty(TranslateComponent.prototype, "key", {
        set: function (key) {
            this._key = key;
            this.startTranslation();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TranslateComponent.prototype, "params", {
        set: function (params) {
            if (typeof params !== "object") {
                this.logHandler.error("Params have to be an object");
                return;
            }
            this._params = params;
            this.startTranslation();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TranslateComponent.prototype, "module", {
        set: function (module) {
            var _this = this;
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
            this.translator = this.translatorContainer.getTranslator(module);
            this.subscription = this.translator.languageChanged.subscribe(function () {
                _this.startTranslation();
            });
            this.startTranslation();
        },
        enumerable: true,
        configurable: true
    });
    TranslateComponent.prototype.startTranslation = function () {
        var _this = this;
        if (!this._key) {
            return;
        }
        this.translator.translate(this._key, this._params).then(function (translation) { return _this.translation = String(translation); });
    };
    __decorate([
        Input("translate"),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TranslateComponent.prototype, "key", null);
    TranslateComponent = __decorate([
        Component({
            selector: "[translate]",
            template: "{{translation}}",
            inputs: [
                "params:translateParams",
                "module:translatorModule",
            ],
        }),
        __metadata("design:paramtypes", [Translator,
            TranslatorContainer,
            TranslateLogHandler])
    ], TranslateComponent);
    return TranslateComponent;
}());
export { TranslateComponent };
