// ngx-dropzone/bundles/ngx-dropzone.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-dropzone', ['exports', '@angular/core', '@angular/platform-browser', '@angular/common'], factory) :
    (factory((global['ngx-dropzone'] = {}),global.ng.core,global.ng.platformBrowser,global.ng.common));
}(this, (function (exports,core,platformBrowser,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneLabelDirective = /** @class */ (function () {
        function NgxDropzoneLabelDirective() {
        }
        NgxDropzoneLabelDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'ngx-dropzone-label'
                    },] },
        ];
        return NgxDropzoneLabelDirective;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Coerces a data-bound value (typically a string) to a boolean.
     * Taken from https://github.com/angular/components/blob/master/src/cdk/coercion/boolean-property.ts
     * @param {?} value
     * @return {?}
     */
    function coerceBooleanProperty(value) {
        return value != null && "" + value !== 'false';
    }
    /**
     * Whether the provided value is considered a number.
     * Taken from https://github.com/angular/components/blob/master/src/cdk/coercion/number-property.ts
     * @param {?} value
     * @return {?}
     */
    function coerceNumberProperty(value) {
        // parseFloat(value) handles most of the cases we're interested in (it treats null, empty string,
        // and other non-number values as NaN, where Number just uses 0) but it considers the string
        // '123hello' to be a valid number. Therefore we also check if Number(value) is NaN.
        return (!isNaN(parseFloat(( /** @type {?} */(value)))) && !isNaN(Number(value))) ? Number(value) : null;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {number} */
    var KEY_CODE = {
        BACKSPACE: 8,
        DELETE: 46,
    };
    KEY_CODE[KEY_CODE.BACKSPACE] = 'BACKSPACE';
    KEY_CODE[KEY_CODE.DELETE] = 'DELETE';
    var NgxDropzonePreviewComponent = /** @class */ (function () {
        function NgxDropzonePreviewComponent(sanitizer) {
            this.sanitizer = sanitizer;
            this._removable = false;
            /**
             * Emitted when the element should be removed.
             */
            this.removed = new core.EventEmitter();
            /**
             * Make the preview item focusable using the tab key.
             */
            this.tabIndex = 0;
        }
        Object.defineProperty(NgxDropzonePreviewComponent.prototype, "removable", {
            /** Allow the user to remove files. */
            get: /**
             * Allow the user to remove files.
             * @return {?}
             */ function () {
                return this._removable;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._removable = coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzonePreviewComponent.prototype.keyEvent = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                switch (event.keyCode) {
                    case KEY_CODE.BACKSPACE:
                    case KEY_CODE.DELETE:
                        this.remove();
                        break;
                    default:
                        break;
                }
            };
        Object.defineProperty(NgxDropzonePreviewComponent.prototype, "hostStyle", {
            /** We use the HostBinding to pass these common styles to child components. */
            get: /**
             * We use the HostBinding to pass these common styles to child components.
             * @return {?}
             */ function () {
                /** @type {?} */
                var styles = "\n\t\t\tdisplay: flex;\n\t\t\theight: 140px;\n\t\t\tmin-height: 140px;\n\t\t\tmin-width: 180px;\n\t\t\tmax-width: 180px;\n\t\t\tjustify-content: center;\n\t\t\talign-items: center;\n\t\t\tpadding: 0 20px;\n\t\t\tmargin: 10px;\n\t\t\tborder-radius: 5px;\n\t\t\tposition: relative;\n\t\t";
                return this.sanitizer.bypassSecurityTrustStyle(styles);
            },
            enumerable: true,
            configurable: true
        });
        /** Remove method to be used from the template. */
        /**
         * Remove method to be used from the template.
         * @param {?} event
         * @return {?}
         */
        NgxDropzonePreviewComponent.prototype._remove = /**
         * Remove method to be used from the template.
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.stopPropagation();
                this.remove();
            };
        /** Remove the preview item (use from component code). */
        /**
         * Remove the preview item (use from component code).
         * @return {?}
         */
        NgxDropzonePreviewComponent.prototype.remove = /**
         * Remove the preview item (use from component code).
         * @return {?}
         */
            function () {
                if (this._removable) {
                    this.removed.next(this.file);
                }
            };
        /**
         * @protected
         * @return {?}
         */
        NgxDropzonePreviewComponent.prototype.readFile = /**
         * @protected
         * @return {?}
         */
            function () {
                return __awaiter(this, void 0, void 0, function () {
                    var _this = this;
                    return __generator(this, function (_a) {
                        return [2 /*return*/, new Promise(( /**
                                 * @param {?} resolve
                                 * @param {?} reject
                                 * @return {?}
                                 */function (resolve, reject) {
                                /** @type {?} */
                                var reader = new FileReader();
                                reader.onload = ( /**
                                 * @param {?} e
                                 * @return {?}
                                 */function (e) {
                                    resolve((( /** @type {?} */(e.target))).result);
                                });
                                reader.onerror = ( /**
                                 * @param {?} e
                                 * @return {?}
                                 */function (e) {
                                    console.error("FileReader failed on file " + _this.file.name + ".");
                                    reject(e);
                                });
                                if (!_this.file) {
                                    return reject('No file to read. Please provide a file using the [file] Input property.');
                                }
                                reader.readAsDataURL(_this.file);
                            }))];
                    });
                });
            };
        NgxDropzonePreviewComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone-preview',
                        template: "\n\t\t<ng-content select=\"ngx-dropzone-label\"></ng-content>\n\t\t<ngx-dropzone-remove-badge *ngIf=\"removable\" (click)=\"_remove($event)\">\n\t\t</ngx-dropzone-remove-badge>\n\t",
                        styles: [":host(){background-image:linear-gradient(to top,#ededed,#efefef,#f1f1f1,#f4f4f4,#f6f6f6)}:host():focus,:host():hover{background-image:linear-gradient(to top,#e3e3e3,#ebeaea,#e8e7e7,#ebeaea,#f4f4f4);outline:0}:host():focus ngx-dropzone-remove-badge,:host():hover ngx-dropzone-remove-badge{opacity:1}:host() ngx-dropzone-remove-badge{opacity:0}:host() ::ng-deep ngx-dropzone-label{overflow-wrap:break-word}"]
                    },] },
        ];
        /** @nocollapse */
        NgxDropzonePreviewComponent.ctorParameters = function () {
            return [
                { type: platformBrowser.DomSanitizer }
            ];
        };
        NgxDropzonePreviewComponent.propDecorators = {
            file: [{ type: core.Input }],
            removable: [{ type: core.Input }],
            removed: [{ type: core.Output }],
            keyEvent: [{ type: core.HostListener, args: ['keyup', ['$event'],] }],
            hostStyle: [{ type: core.HostBinding, args: ['style',] }],
            tabIndex: [{ type: core.HostBinding, args: ['tabindex',] }]
        };
        return NgxDropzonePreviewComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * This service contains the filtering logic to be applied to
     * any dropped or selected file. If a file matches all criteria
     * like maximum size or accept type, it will be emitted in the
     * addedFiles array, otherwise in the rejectedFiles array.
     */
    var NgxDropzoneService = /** @class */ (function () {
        function NgxDropzoneService() {
        }
        /**
         * @param {?} files
         * @param {?} accept
         * @param {?} maxFileSize
         * @param {?} multiple
         * @return {?}
         */
        NgxDropzoneService.prototype.parseFileList = /**
         * @param {?} files
         * @param {?} accept
         * @param {?} maxFileSize
         * @param {?} multiple
         * @return {?}
         */
            function (files, accept, maxFileSize, multiple) {
                /** @type {?} */
                var addedFiles = [];
                /** @type {?} */
                var rejectedFiles = [];
                for (var i = 0; i < files.length; i++) {
                    /** @type {?} */
                    var file = files.item(i);
                    if (!this.isAccepted(file, accept)) {
                        this.rejectFile(rejectedFiles, file, 'type');
                        continue;
                    }
                    if (maxFileSize && file.size > maxFileSize) {
                        this.rejectFile(rejectedFiles, file, 'size');
                        continue;
                    }
                    if (!multiple && addedFiles.length >= 1) {
                        this.rejectFile(rejectedFiles, file, 'no_multiple');
                        continue;
                    }
                    addedFiles.push(file);
                }
                /** @type {?} */
                var result = {
                    addedFiles: addedFiles,
                    rejectedFiles: rejectedFiles
                };
                return result;
            };
        /**
         * @private
         * @param {?} file
         * @param {?} accept
         * @return {?}
         */
        NgxDropzoneService.prototype.isAccepted = /**
         * @private
         * @param {?} file
         * @param {?} accept
         * @return {?}
         */
            function (file, accept) {
                if (accept === '*') {
                    return true;
                }
                /** @type {?} */
                var acceptFiletypes = accept.split(',').map(( /**
                 * @param {?} it
                 * @return {?}
                 */function (it) { return it.toLowerCase().trim(); }));
                /** @type {?} */
                var filetype = file.type.toLowerCase();
                /** @type {?} */
                var filename = file.name.toLowerCase();
                /** @type {?} */
                var matchedFileType = acceptFiletypes.find(( /**
                 * @param {?} acceptFiletype
                 * @return {?}
                 */function (acceptFiletype) {
                    // check for wildcard mimetype (e.g. image/*)
                    if (acceptFiletype.endsWith('/*')) {
                        return filetype.split('/')[0] === acceptFiletype.split('/')[0];
                    }
                    // check for file extension (e.g. .csv)
                    if (acceptFiletype.startsWith(".")) {
                        return filename.endsWith(acceptFiletype);
                    }
                    // check for exact mimetype match (e.g. image/jpeg)
                    return acceptFiletype == filetype;
                }));
                return !!matchedFileType;
            };
        /**
         * @private
         * @param {?} rejectedFiles
         * @param {?} file
         * @param {?} reason
         * @return {?}
         */
        NgxDropzoneService.prototype.rejectFile = /**
         * @private
         * @param {?} rejectedFiles
         * @param {?} file
         * @param {?} reason
         * @return {?}
         */
            function (rejectedFiles, file, reason) {
                /** @type {?} */
                var rejectedFile = ( /** @type {?} */(file));
                rejectedFile.reason = reason;
                rejectedFiles.push(rejectedFile);
            };
        NgxDropzoneService.decorators = [
            { type: core.Injectable },
        ];
        return NgxDropzoneService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneComponent = /** @class */ (function () {
        function NgxDropzoneComponent(service) {
            this.service = service;
            /**
             * Emitted when any files were added or rejected.
             */
            this.change = new core.EventEmitter();
            /**
             * Set the accepted file types. Defaults to '*'.
             */
            this.accept = '*';
            this._disabled = false;
            this._multiple = true;
            this._maxFileSize = undefined;
            this._expandable = false;
            this._disableClick = false;
            this._isHovered = false;
        }
        Object.defineProperty(NgxDropzoneComponent.prototype, "_hasPreviews", {
            get: /**
             * @return {?}
             */ function () {
                return !!this._previewChildren.length;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxDropzoneComponent.prototype, "disabled", {
            /** Disable any user interaction with the component. */
            get: /**
             * Disable any user interaction with the component.
             * @return {?}
             */ function () {
                return this._disabled;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._disabled = coerceBooleanProperty(value);
                if (this._isHovered) {
                    this._isHovered = false;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxDropzoneComponent.prototype, "multiple", {
            /** Allow the selection of multiple files. */
            get: /**
             * Allow the selection of multiple files.
             * @return {?}
             */ function () {
                return this._multiple;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._multiple = coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxDropzoneComponent.prototype, "maxFileSize", {
            /** Set the maximum size a single file may have. */
            get: /**
             * Set the maximum size a single file may have.
             * @return {?}
             */ function () {
                return this._maxFileSize;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._maxFileSize = coerceNumberProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxDropzoneComponent.prototype, "expandable", {
            /** Allow the dropzone container to expand vertically. */
            get: /**
             * Allow the dropzone container to expand vertically.
             * @return {?}
             */ function () {
                return this._expandable;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._expandable = coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxDropzoneComponent.prototype, "disableClick", {
            /** Open the file selector on click. */
            get: /**
             * Open the file selector on click.
             * @return {?}
             */ function () {
                return this._disableClick;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._disableClick = coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        /** Show the native OS file explorer to select files. */
        /**
         * Show the native OS file explorer to select files.
         * @return {?}
         */
        NgxDropzoneComponent.prototype._onClick = /**
         * Show the native OS file explorer to select files.
         * @return {?}
         */
            function () {
                if (!this.disableClick) {
                    this.showFileSelector();
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype._onDragOver = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.disabled) {
                    return;
                }
                this.preventDefault(event);
                this._isHovered = true;
            };
        /**
         * @return {?}
         */
        NgxDropzoneComponent.prototype._onDragLeave = /**
         * @return {?}
         */
            function () {
                this._isHovered = false;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype._onDrop = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.disabled) {
                    return;
                }
                this.preventDefault(event);
                this._isHovered = false;
                this.handleFileDrop(event.dataTransfer.files);
            };
        /**
         * @return {?}
         */
        NgxDropzoneComponent.prototype.showFileSelector = /**
         * @return {?}
         */
            function () {
                if (!this.disabled) {
                    (( /** @type {?} */(this._fileInput.nativeElement))).click();
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype._onFilesSelected = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var files = event.target.files;
                this.handleFileDrop(files);
                // Reset the native file input element to allow selecting the same file again
                this._fileInput.nativeElement.value = '';
                // fix(#32): Prevent the default event behaviour which caused the change event to emit twice.
                this.preventDefault(event);
            };
        /**
         * @private
         * @param {?} files
         * @return {?}
         */
        NgxDropzoneComponent.prototype.handleFileDrop = /**
         * @private
         * @param {?} files
         * @return {?}
         */
            function (files) {
                /** @type {?} */
                var result = this.service.parseFileList(files, this.accept, this.maxFileSize, this.multiple);
                this.change.next({
                    addedFiles: result.addedFiles,
                    rejectedFiles: result.rejectedFiles,
                    source: this
                });
            };
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.preventDefault = /**
         * @private
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                event.stopPropagation();
            };
        NgxDropzoneComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone, [ngx-dropzone]',
                        template: "<input #fileInput type=\"file\" [id]=\"id\" [multiple]=\"multiple\" [accept]=\"accept\" [disabled]=\"disabled\"\n  (change)=\"_onFilesSelected($event)\" [attr.aria-label]=\"ariaLabel\" [attr.aria-labelledby]=\"ariaLabelledby\"\n  [attr.aria-describedby]=\"ariaDescribedBy\">\n<ng-content select=\"ngx-dropzone-label\" *ngIf=\"!_hasPreviews\"></ng-content>\n<ng-content select=\"ngx-dropzone-preview\"></ng-content>\n<ng-content></ng-content>\n",
                        styles: [":host(){display:flex;align-items:center;height:180px;background:#fff;cursor:pointer;color:#717386;border:2px dashed #717386;border-radius:5px;font-size:16px;overflow-x:auto}:host().ngx-dz-hovered{border-style:solid}:host().ngx-dz-disabled{opacity:.5;cursor:no-drop;pointer-events:none}:host().expandable{overflow:hidden;height:unset;min-height:180px;flex-wrap:wrap}:host().unclickable{cursor:default}:host() ::ng-deep ngx-dropzone-label{text-align:center;z-index:10;margin:10px auto}:host() input{width:.1px;height:.1px;opacity:0;overflow:hidden;position:absolute;z-index:-1}:host() input:focus+::ng-deep ngx-dropzone-label{outline:#000 dotted 1px;outline:-webkit-focus-ring-color auto 5px}"],
                        providers: [NgxDropzoneService]
                    },] },
        ];
        /** @nocollapse */
        NgxDropzoneComponent.ctorParameters = function () {
            return [
                { type: NgxDropzoneService, decorators: [{ type: core.Self }] }
            ];
        };
        NgxDropzoneComponent.propDecorators = {
            _previewChildren: [{ type: core.ContentChildren, args: [NgxDropzonePreviewComponent, { descendants: true },] }],
            _fileInput: [{ type: core.ViewChild, args: ['fileInput',] }],
            change: [{ type: core.Output }],
            accept: [{ type: core.Input }],
            disabled: [{ type: core.Input }, { type: core.HostBinding, args: ['class.ngx-dz-disabled',] }],
            multiple: [{ type: core.Input }],
            maxFileSize: [{ type: core.Input }],
            expandable: [{ type: core.Input }, { type: core.HostBinding, args: ['class.expandable',] }],
            disableClick: [{ type: core.Input }, { type: core.HostBinding, args: ['class.unclickable',] }],
            id: [{ type: core.Input }],
            ariaLabel: [{ type: core.Input, args: ['aria-label',] }],
            ariaLabelledby: [{ type: core.Input, args: ['aria-labelledby',] }],
            ariaDescribedBy: [{ type: core.Input, args: ['aria-describedby',] }],
            _isHovered: [{ type: core.HostBinding, args: ['class.ngx-dz-hovered',] }],
            _onClick: [{ type: core.HostListener, args: ['click',] }],
            _onDragOver: [{ type: core.HostListener, args: ['dragover', ['$event'],] }],
            _onDragLeave: [{ type: core.HostListener, args: ['dragleave',] }],
            _onDrop: [{ type: core.HostListener, args: ['drop', ['$event'],] }]
        };
        return NgxDropzoneComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneImagePreviewComponent = /** @class */ (function (_super) {
        __extends(NgxDropzoneImagePreviewComponent, _super);
        function NgxDropzoneImagePreviewComponent(sanitizer) {
            var _this = _super.call(this, sanitizer) || this;
            /**
             * The image data source.
             */
            _this.imageSrc = '';
            return _this;
        }
        /**
         * @return {?}
         */
        NgxDropzoneImagePreviewComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.readFile()
                    .then(( /**
             * @param {?} img
             * @return {?}
             */function (img) {
                    return setTimeout(( /**
                     * @return {?}
                     */function () { return _this.imageSrc = img; }));
                }))
                    .catch(( /**
             * @param {?} err
             * @return {?}
             */function (err) { return console.error(err); }));
            };
        NgxDropzoneImagePreviewComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone-image-preview',
                        template: "\n    <img [src]=\"imageSrc\" />\n\t\t<ng-content select=\"ngx-dropzone-label\"></ng-content>\n    <ngx-dropzone-remove-badge *ngIf=\"removable\" (click)=\"_remove($event)\">\n    </ngx-dropzone-remove-badge>\n\t",
                        styles: [":host(){min-width:unset!important;max-width:unset!important;padding:0!important}:host():focus img,:host():hover img{opacity:.7}:host():focus ngx-dropzone-remove-badge,:host():hover ngx-dropzone-remove-badge{opacity:1}:host() ngx-dropzone-remove-badge{opacity:0}:host() img{max-height:100%;border-radius:5px;opacity:.8}:host() ::ng-deep ngx-dropzone-label{position:absolute;overflow-wrap:break-word}"],
                        providers: [
                            {
                                provide: NgxDropzonePreviewComponent,
                                useExisting: NgxDropzoneImagePreviewComponent
                            }
                        ]
                    },] },
        ];
        /** @nocollapse */
        NgxDropzoneImagePreviewComponent.ctorParameters = function () {
            return [
                { type: platformBrowser.DomSanitizer }
            ];
        };
        return NgxDropzoneImagePreviewComponent;
    }(NgxDropzonePreviewComponent));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneRemoveBadgeComponent = /** @class */ (function () {
        function NgxDropzoneRemoveBadgeComponent() {
        }
        NgxDropzoneRemoveBadgeComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone-remove-badge',
                        template: "\n    <svg>\n      <line x1=\"0\" y1=\"0\" x2=\"10\" y2=\"10\" />\n      <line x1=\"0\" y1=\"10\" x2=\"10\" y2=\"0\" />\n    </svg>\n  ",
                        styles: [":host(){display:flex;justify-content:center;align-items:center;height:22px;width:22px;position:absolute;top:5px;right:5px;border-radius:50%;background:#bbb;color:#333;cursor:pointer}:host():hover{background:#aeaeae}:host()>svg{height:10px;width:10px}:host()>svg>line{stroke-width:2px;stroke:#fff}"]
                    },] },
        ];
        return NgxDropzoneRemoveBadgeComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneVideoPreviewComponent = /** @class */ (function (_super) {
        __extends(NgxDropzoneVideoPreviewComponent, _super);
        function NgxDropzoneVideoPreviewComponent(sanitizer) {
            return _super.call(this, sanitizer) || this;
        }
        /**
         * @return {?}
         */
        NgxDropzoneVideoPreviewComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                if (!this.file) {
                    console.error('No file to read. Please provide a file using the [file] Input property.');
                    return;
                }
                /**
                 * We sanitize the URL here to enable the preview.
                 * Please note that this could cause security issues!
                 **/
                this.videoSrc = URL.createObjectURL(this.file);
                this.sanitizedVideoSrc = this.sanitizer.bypassSecurityTrustUrl(this.videoSrc);
            };
        /**
         * @return {?}
         */
        NgxDropzoneVideoPreviewComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                URL.revokeObjectURL(this.videoSrc);
            };
        NgxDropzoneVideoPreviewComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone-video-preview',
                        template: "\n    <video *ngIf=\"sanitizedVideoSrc\" controls (click)=\"$event.stopPropagation()\">\n      <source [src]=\"sanitizedVideoSrc\" />\n    </video>\n    <ng-content select=\"ngx-dropzone-label\"></ng-content>\n    <ngx-dropzone-remove-badge *ngIf=\"removable\" (click)=\"_remove($event)\">\n    </ngx-dropzone-remove-badge>\n\t",
                        styles: [":host(){min-width:unset!important;max-width:unset!important;padding:0!important}:host():focus video,:host():hover video{opacity:.7}:host():focus ngx-dropzone-remove-badge,:host():hover ngx-dropzone-remove-badge{opacity:1}:host() ngx-dropzone-remove-badge{opacity:0}:host() video{max-height:100%;border-radius:5px}:host() ::ng-deep ngx-dropzone-label{position:absolute;overflow-wrap:break-word}"],
                        providers: [
                            {
                                provide: NgxDropzonePreviewComponent,
                                useExisting: NgxDropzoneVideoPreviewComponent
                            }
                        ]
                    },] },
        ];
        /** @nocollapse */
        NgxDropzoneVideoPreviewComponent.ctorParameters = function () {
            return [
                { type: platformBrowser.DomSanitizer }
            ];
        };
        return NgxDropzoneVideoPreviewComponent;
    }(NgxDropzonePreviewComponent));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneModule = /** @class */ (function () {
        function NgxDropzoneModule() {
        }
        NgxDropzoneModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            NgxDropzoneComponent,
                            NgxDropzoneLabelDirective,
                            NgxDropzonePreviewComponent,
                            NgxDropzoneImagePreviewComponent,
                            NgxDropzoneRemoveBadgeComponent,
                            NgxDropzoneVideoPreviewComponent,
                        ],
                        exports: [
                            NgxDropzoneComponent,
                            NgxDropzoneLabelDirective,
                            NgxDropzonePreviewComponent,
                            NgxDropzoneImagePreviewComponent,
                            NgxDropzoneRemoveBadgeComponent,
                            NgxDropzoneVideoPreviewComponent,
                        ]
                    },] },
        ];
        return NgxDropzoneModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NgxDropzoneModule = NgxDropzoneModule;
    exports.NgxDropzoneComponent = NgxDropzoneComponent;
    exports.NgxDropzonePreviewComponent = NgxDropzonePreviewComponent;
    exports.NgxDropzoneImagePreviewComponent = NgxDropzoneImagePreviewComponent;
    exports.NgxDropzoneVideoPreviewComponent = NgxDropzoneVideoPreviewComponent;
    exports.NgxDropzoneRemoveBadgeComponent = NgxDropzoneRemoveBadgeComponent;
    exports.ɵb = NgxDropzoneLabelDirective;
    exports.ɵa = NgxDropzoneService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
