// ngx-dropzone/bundles/ngx-dropzone.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-dropzone', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['ngx-dropzone'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * UPDATE 04.04.2019:
     * Refactored to use service class to handle any
     * logic on the dropped files to allow for easier
     * unit tests and separation of concerns.
     */
    var NgxDropzoneService = /** @class */ (function () {
        function NgxDropzoneService() {
            this.fileCache = [];
            this.rejectedFiles = [];
            this.previews = [];
        }
        /**
         * @return {?}
         */
        NgxDropzoneService.prototype.reset = /**
         * @return {?}
         */
            function () {
                this.fileCache = [];
                this.rejectedFiles = [];
                this.previews = [];
            };
        /**
         * @param {?} files
         * @param {?} accept
         * @param {?} maxFileSize
         * @param {?} multiple
         * @param {?} preserveFiles
         * @param {?} showPreviews
         * @return {?}
         */
        NgxDropzoneService.prototype.parseFileList = /**
         * @param {?} files
         * @param {?} accept
         * @param {?} maxFileSize
         * @param {?} multiple
         * @param {?} preserveFiles
         * @param {?} showPreviews
         * @return {?}
         */
            function (files, accept, maxFileSize, multiple, preserveFiles, showPreviews) {
                return __awaiter(this, void 0, void 0, function () {
                    var hasFiletypeFilter, i, file, preview, preview, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                hasFiletypeFilter = accept !== '*';
                                /**
                                 * UPDATE 12.03.2019:
                                 * Added option to preserve preview images.
                                 */
                                if (!preserveFiles) {
                                    this.fileCache = [];
                                    this.rejectedFiles = [];
                                    this.previews = [];
                                }
                                i = 0;
                                _a.label = 1;
                            case 1:
                                if (!(i < files.length))
                                    return [3 /*break*/, 6];
                                file = files.item(i);
                                if (hasFiletypeFilter) {
                                    if (accept.endsWith('/*')) {
                                        // If a generic file type is provided, we check for a match.
                                        if (accept.split('/')[0] !== file.type.split('/')[0]) {
                                            this.rejectedFiles.push(file);
                                            return [3 /*break*/, 5];
                                        }
                                    }
                                    else {
                                        // Else an exact match is required.
                                        if (!accept.includes(file.type)) {
                                            this.rejectedFiles.push(file);
                                            return [3 /*break*/, 5];
                                        }
                                    }
                                }
                                if (maxFileSize && file.size > maxFileSize) {
                                    this.rejectedFiles.push(file);
                                    return [3 /*break*/, 5];
                                }
                                if (!multiple && this.fileCache.length >= 1) {
                                    if (!preserveFiles) {
                                        // Always emit the latest file if multi-selection and preservation are disabled.
                                        this.fileCache = [file];
                                    }
                                    else {
                                        return [3 /*break*/, 5];
                                    }
                                }
                                if (!showPreviews)
                                    return [3 /*break*/, 4];
                                if (!file.type.startsWith('image'))
                                    return [3 /*break*/, 3];
                                return [4 /*yield*/, this.readFile(file)];
                            case 2:
                                preview = _a.sent();
                                if (preview) {
                                    this.previews.push(preview);
                                }
                                return [3 /*break*/, 4];
                            case 3:
                                preview = {
                                    data: null,
                                    filename: file.name
                                };
                                this.previews.push(preview);
                                _a.label = 4;
                            case 4:
                                this.fileCache.push(file);
                                _a.label = 5;
                            case 5:
                                i++;
                                return [3 /*break*/, 1];
                            case 6:
                                result = {
                                    addedFiles: this.fileCache,
                                    rejectedFiles: this.rejectedFiles
                                };
                                return [2 /*return*/, result];
                        }
                    });
                });
            };
        /**
         * @private
         * @param {?} file
         * @return {?}
         */
        NgxDropzoneService.prototype.readFile = /**
         * @private
         * @param {?} file
         * @return {?}
         */
            function (file) {
                return __awaiter(this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        return [2 /*return*/, new Promise(( /**
                                 * @param {?} resolve
                                 * @param {?} reject
                                 * @return {?}
                                 */function (resolve, reject) {
                                /** @type {?} */
                                var reader = new FileReader();
                                reader.onload = ( /**
                                 * @param {?} e
                                 * @return {?}
                                 */function (e) {
                                    /** @type {?} */
                                    var preview = {
                                        data: (( /** @type {?} */(e.target))).result,
                                        filename: file.name
                                    };
                                    return resolve(preview);
                                });
                                reader.onerror = ( /**
                                 * @param {?} e
                                 * @return {?}
                                 */function (e) {
                                    console.error("FileReader failed on file " + file.name + ". No preview image created.");
                                    return reject(null);
                                });
                                reader.readAsDataURL(file);
                            }))];
                    });
                });
            };
        NgxDropzoneService.decorators = [
            { type: core.Injectable },
        ];
        /** @nocollapse */
        NgxDropzoneService.ctorParameters = function () { return []; };
        return NgxDropzoneService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneComponent = /** @class */ (function () {
        function NgxDropzoneComponent(host, service) {
            this.host = host;
            this.service = service;
            this.label = 'Drop your files here (or click)';
            this.multiple = true;
            this.accept = '*';
            this.showPreviews = false;
            this.preserveFiles = true;
            this.filesAdded = new core.EventEmitter();
            this.filesRejected = new core.EventEmitter();
            this.disabled = false;
            this.hovered = false;
        }
        /**
         * @return {?}
         */
        NgxDropzoneComponent.prototype.showFileSelector = /**
         * @return {?}
         */
            function () {
                if (!this.disabled) {
                    this.fileInput.nativeElement.click();
                }
            };
        /**
         * @return {?}
         */
        NgxDropzoneComponent.prototype.reset = /**
         * @return {?}
         */
            function () {
                this.service.reset();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.onFilesSelected = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                var _this = this;
                /** @type {?} */
                var files = event.target.files;
                this.handleFileDrop(files).then(( /**
                 * @return {?}
                 */function () {
                    // Reset the file input value to trigger the event on new selection.
                    (( /** @type {?} */(_this.fileInput.nativeElement))).value = '';
                }));
            };
        /**
         * UPDATE 10.03.2019:
         * Refactored to use HostListener and HostBindings to allow
         * for easier style overwriting from outside the component.
         */
        /**
         * UPDATE 10.03.2019:
         * Refactored to use HostListener and HostBindings to allow
         * for easier style overwriting from outside the component.
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.onDragOver = /**
         * UPDATE 10.03.2019:
         * Refactored to use HostListener and HostBindings to allow
         * for easier style overwriting from outside the component.
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.disabled) {
                    return;
                }
                this.preventDefault(event);
                this.hovered = true;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.onDragLeave = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.hovered = false;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.onDrop = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.preventDefault(event);
                this.hovered = false;
                this.handleFileDrop(event.dataTransfer.files);
            };
        /**
         * @private
         * @param {?} files
         * @return {?}
         */
        NgxDropzoneComponent.prototype.handleFileDrop = /**
         * @private
         * @param {?} files
         * @return {?}
         */
            function (files) {
                return __awaiter(this, void 0, void 0, function () {
                    var _this = this;
                    return __generator(this, function (_a) {
                        return [2 /*return*/, new Promise(( /**
                                 * @param {?} resolve
                                 * @return {?}
                                 */function (resolve) {
                                if (_this.disabled) {
                                    return;
                                }
                                _this.service.parseFileList(files, _this.accept, _this.maxFileSize, _this.multiple, _this.preserveFiles, _this.showPreviews)
                                    .then(( /**
                             * @param {?} result
                             * @return {?}
                             */function (result) {
                                    _this.filesAdded.next(result.addedFiles);
                                    if (result.rejectedFiles.length) {
                                        _this.filesRejected.next(result.rejectedFiles);
                                    }
                                    resolve();
                                }));
                            }))];
                    });
                });
            };
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        NgxDropzoneComponent.prototype.preventDefault = /**
         * @private
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                event.stopPropagation();
            };
        NgxDropzoneComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-dropzone',
                        template: "<input type=\"file\" #fileInput class=\"file-input\" (change)=\"onFilesSelected($event)\" [multiple]=\"multiple\"\n  [accept]=\"accept\" />\n<div class=\"dropzone\" (click)=\"showFileSelector()\">\n  <p *ngIf=\"service.previews.length === 0\">{{ label }}</p>\n  <div class=\"preview\" [class.limit-width]=\"!p.data\" *ngFor=\"let p of service.previews\">\n    <img *ngIf=\"p.data\" [src]=\"p.data\">\n    <span *ngIf=\"p.data\">{{ p.filename }}</span>\n    <div *ngIf=\"!p.data\" class=\"preview-item\">\n      <span>{{ p.filename }}</span>\n    </div>\n  </div>\n</div>\n",
                        styles: [":host(){display:flex;height:180px;cursor:pointer;background:#fff;color:#717386;border:2px dashed #717386;border-radius:5px;font-size:16px}:host().hovered{border:2px solid #717386;color:#dfdfe4}:host().disabled{opacity:.5;cursor:no-drop}:host()>.dropzone{display:flex;flex-direction:row;justify-content:flex-start;align-items:center;width:100%;overflow-x:auto}:host()>.dropzone>p{margin:10px auto}:host()>.dropzone>.preview{height:80%;position:relative;text-align:center;padding:0 10px;width:-webkit-max-content;width:-moz-max-content;width:max-content}:host()>.dropzone>.preview.limit-width{max-width:25%}:host()>.dropzone>.preview>img{max-height:100%;border-radius:5px;opacity:.8}:host()>.dropzone>.preview>span{position:absolute;top:50%;left:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);color:#fff}:host()>.dropzone>.preview>.preview-item{display:flex;text-align:center;align-items:center;height:100%;padding:0 40px;overflow-wrap:break-word;border-radius:5px;background-image:linear-gradient(to top,#ededed,#efefef,#f1f1f1,#f4f4f4,#f6f6f6)}:host()>.dropzone>.preview>.preview-item>span{color:#717386;max-width:100%}.file-input{display:none}"],
                        providers: [NgxDropzoneService] // Create a new service instance for each component.
                    },] },
        ];
        /** @nocollapse */
        NgxDropzoneComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef },
                { type: NgxDropzoneService }
            ];
        };
        NgxDropzoneComponent.propDecorators = {
            label: [{ type: core.Input }],
            multiple: [{ type: core.Input }],
            accept: [{ type: core.Input }],
            maxFileSize: [{ type: core.Input }],
            showPreviews: [{ type: core.Input }],
            preserveFiles: [{ type: core.Input }],
            filesAdded: [{ type: core.Output }],
            filesRejected: [{ type: core.Output }],
            disabled: [{ type: core.HostBinding, args: ['class.disabled',] }, { type: core.Input }],
            hovered: [{ type: core.HostBinding, args: ['class.hovered',] }],
            fileInput: [{ type: core.ViewChild, args: ['fileInput',] }],
            onDragOver: [{ type: core.HostListener, args: ['dragover', ['$event'],] }],
            onDragLeave: [{ type: core.HostListener, args: ['dragleave', ['$event'],] }],
            onDrop: [{ type: core.HostListener, args: ['drop', ['$event'],] }]
        };
        return NgxDropzoneComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxDropzoneModule = /** @class */ (function () {
        function NgxDropzoneModule() {
        }
        NgxDropzoneModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            NgxDropzoneComponent
                        ],
                        exports: [
                            NgxDropzoneComponent
                        ]
                    },] },
        ];
        return NgxDropzoneModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NgxDropzoneComponent = NgxDropzoneComponent;
    exports.NgxDropzoneModule = NgxDropzoneModule;
    exports.ɵa = NgxDropzoneService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
